samcli/lib/iac/cfn/cfn_iac.py

```diff
@@ -1,33 +1,172 @@
         the CFN project type
     """
 
+    def __init__(self, context: SamCliContext):
+        self._template_file = context.command_options_map["template_file"]
+        self._base_dir = context.command_options_map.get("base_dir", None)
+        super().__init__(context)
+
     def read_project(self, lookup_paths: List[LookupPath]) -> SamCliProject:
-        pass
+        stacks = [self._build_stack(self._template_file)]
+
+        return SamCliProject(stacks)
 
     def write_project(self, project: SamCliProject, build_dir: str) -> bool:
+        # TODO
         pass
 
     def update_packaged_locations(self, stack: Stack) -> bool:
+        # TODO
         pass
 
+    # pylint: disable=too-many-branches
+    def _build_stack(self, path: str, is_nested: bool = False, name: Optional[str] = None) -> Stack:
+        assets: List[Asset] = []
+
+        if os.path.islink(path):
+            path = os.path.realpath(path)
+
+        base_dir = self._base_dir or os.path.dirname(path)
+
+        stack = Stack(is_nested=is_nested, name=name, assets=assets, origin_dir=os.path.dirname(path))
+
+        template_dict = get_template_data(path)
+        for key, value in template_dict.items():
+            stack[key] = value
+
+        resources_section = stack.get("Resources", DictSection())
+        for resource in resources_section.section_items:
+            resource_id = resource.item_id
+            resource_type = resource.get("Type", None)
+            properties = resource.get("Properties", {})
+            package_type = properties.get("PackageType", ZIP)
+
+            resource_assets = []
+
+            if resource_type in NESTED_STACKS_RESOURCES:
+                nested_stack = self._extract_nested_stack(path, resource_id, properties, resource_type)
+                resource.nested_stack = nested_stack
+
+            if resource_type in RESOURCES_WITH_LOCAL_PATHS:
+                for path_prop_name in RESOURCES_WITH_LOCAL_PATHS[resource_type]:
+                    asset_path = jmespath.search(path_prop_name, properties)
+                    if is_local_path(asset_path) and package_type == ZIP:
+                        reference_path = base_dir if resource_type in BASE_DIR_RESOURCES else os.path.dirname(path)
+                        asset_path = get_local_path(asset_path, reference_path)
+                        asset = S3Asset(source_path=asset_path, source_property=path_prop_name)
+                        resource_assets.append(asset)
+                        stack.assets.append(asset)
+
+            if resource_type in RESOURCES_WITH_IMAGE_COMPONENT:
+                for path_prop_name in RESOURCES_WITH_IMAGE_COMPONENT[resource_type]:
+                    asset_path = jmespath.search(path_prop_name, properties)
+                    if asset_path and package_type == IMAGE:
+                        asset = ImageAsset(source_local_image=asset_path, source_property=path_prop_name)
```

Comment:
anybody knows how to fix this typing issue? 
`Incompatible types in assignment (expression has type "ImageAsset", variable has type "S3Asset")`
I think mypy infers the type from the first use. I tried pre-declaring the type like this: `asset: Asset` or `asset: Union[S3Asset, ImageAsset]`, but none of these seemed to work.
If we fix this, I think I could remove this file from the section that ignores errors in mypi.ini

Reply:
You can simply rename this one as image_asset, and the one at L115 to s3_asset